### PR TITLE
Implemented new Test::* paradigm

### LINKED ISSUE #25: Broken by Test::Builder alphas
https://github.com/Test-More/test-more/issues/505

This module will break with the new Test::Builder alphas that are already in blead.

The module overrides Test::Builder methods in ways that do not preserve their API, this cannot feasibly be supported. It also directly accesses singleton hash elements that have had proper API's for a long time (test counts).

It is worth noting that the alphas provide api's that would make it possible to implement this module without overriding anything in other modules. If interested ping me and I can provide the basic code for changing the output from Test::*, but I cannot implement the entire module over again for you.

### PR DESCRIPTION
**Pull request by lejeunerenard**
I rewrote `Test::Pretty` to use the new `Test::Stream` API in `Test::More`'s alphas. So far all tests pass with Test::More 1.301001_084. Note that the rewrite is not backwards compatible. If it needs to be, let me know and I'll update the PR.

The code has been structured so that all events emitted by a stream are processed via `stream_listener()`. This way tests can be printed recursively (e.g. subtests). `stream_listener()` adds "pretty" output for 3 different events:
1. `subtest` events are printed as "pretty" `ok()`s if subtests are not set to 'delayed'. Otherwise they print their name and then recurse through their child tests. They do not end with an `ok()` anymore.
2. `ok` events are printed same as before.
3. `plan` events are printed if they are a `SKIP`. The rest are ignored.

Everything else is delegated to it's default rendering.

Closes #25 

_Notes:_
- Added `TEST_PRETTY_INDENT` env variable to set indention
- `subtest` names are no longer half indented.
- To differentiate between a "pretty" ok and the dummy tap, I created a new event called `Test::Stream::Event::DummyTap` which outputs the same dummy tap as before. This utilizes the same filehandle as `Test::More` instead of directly writing to `STDOUT`.
- Removed code that set the exit code so that `Test::More` can do that for us.
- Todo items inside a subtest do not know they are in todo given the context is hidden by `Test::Stream`.

### CODE REVIEW

File: lib/Test/Stream/Event/DummyTap.pm [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package Test::Stream::Event::DummyTap;
```
Comment (tokuhirom) [inline lib/Test/Stream/Event/DummyTap.pm:1]: I guess, if you want to use this name space, you need to discus with @exodist first.

Comment (exodist) [inline lib/Test/Stream/Event/DummyTap.pm:1]: The Test::Stream::Event::\* namespace is intended to be open to additions. I do not foresee any new ones as part of Test-Simple in the near future. If I need one and someone beats me to it I can pick a new name :-)

In other words, +1

File: lib/Test/Stream/Event/DummyTap.pm [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+use strict;
+use warnings;
+
+use Test::Stream::Event(
+   accessors => [qw/succeed/],
+   ctx_method => 'dummy_tap',
+);
+
+sub init {
```
Comment (exodist) [inline lib/Test/Stream/Event/DummyTap.pm]: no need to write your own init

File: lib/Test/Stream/Event/DummyTap.pm [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+sub to_tap {
+   my $self = shift;
+
+   my $ok = ($self->succeed) ? 'ok' : 'not ok';
+
+   return [ OUT_STD, "\n$ok\n", ];
+}
+
+sub extra_details {
```
Comment (exodist) [inline lib/Test/Stream/Event/DummyTap.pm:19]: No need to write this method, though if you want to check the value in Test::Stream::Tester you need to flesh it out more..

**Comment by zoffixznet** [comment]
Hey all. I got this dist assigned to me as part of the [CPAN PR Challenge](http://cpan-prc.org/).

This seems to be a large PR. Is there any way to have it merged before I work through the open Issues?

Travis failure on this PR is due to too low version of the prereq specified.

Let me know. Thanks!

**Comment by lejeunerenard** [comment]
@zoffixznet I believe @tokuhirom hasn't merged the PR because it relies on an alpha version of the Test::More API. I am not sure when the new API is going to be released, but it makes sense to me to not merge this PR into master until then. The PR breaks backwards compatibility.

If you make any changes to master that can be applied to this PR as well, I'd be willing to help integrate them so it's not lost if this PR is merged.

**Comment by zoffixznet** [comment]
Sounds good!

**Comment by preaction** [comment]
Is this ready to go? The Test2 module is out, and #36 and #37 are reported failures to install Test::Pretty. Is there anything I can do to help get this out the door?

**Comment by tokuhirom** [comment]
hmm the test suite is still failing.

**Comment by tokuhirom** [comment]
Travis says following comment. Did you forgot to add the deps?

```
Test/Stream.pm in @INC (@INC contains: /home/travis/build/tokuhirom/Test-Pretty/blib/lib /home/travis/build/tokuhirom/Test-Pretty/blib/arch /home/travis/build/tokuhirom/Test-Pretty/_build/lib /home/travis/perl5/perlbrew/perls/5.16/lib/site_perl/5.16.3/x86_64-linux /home/travis/perl5/perlbrew/perls/5.16/lib/site_perl/5.16.3 /home/travis/perl5/perlbrew/perls/5.16/lib/5.16.3/x86_64-linux /home/travis/perl5/perlbrew/perls/5.16/lib/5.16.3 .) at /home/travis/build/tokuhirom/Test-Pretty/blib/lib/Test/Pretty.pm
```

**Comment by exodist** [comment]
You need to use Test2::API not Test::Stream.

**Comment by preaction** [comment]
I can take a shot at making that change unless someone with more experience has some time

**Comment by lejeunerenard** [comment]
@preaction feel free to tackle this. I haven't touched `Test2` since it was `Test::Stream`.

**Comment by karupanerura** [comment]
How is the progress?

**Comment by tokuhirom** [comment]
@exodist how do i use Test2::API?

**Comment by exodist** [comment]
I think at this point the best thing to do is write a new formatter. Look at https://metacpan.org/source/EXODIST/Test-Simple-1.302067/lib/Test2/Formatter/TAP.pm  as an example implementation.

All you, or anyone else would have to do is load Test2::Formatter::YourFormatter, and it would do the job.

The formatter receives events. Each event will implement the following methods to help you decide what to display (Taken from https://metacpan.org/source/EXODIST/Test-Simple-1.302067/lib/Test2/Event.pm)

```perl
sub causes_fail      { return $bool }       # not ok
sub increments_count { return $bool } # Bumps the test count
sub diagnostics      { return $bool }      # Is a diagnostics message (STDERR)
sub no_display       { return $bool }      # Should not be displayed by most formatters

# Plan details: $max is the count, #directive is something like 'skip all' if set, and 'reason' is the reason to skip all
sub sets_plan { return ($max, $directive, $reason) }
 
# Basically an $event->to_string that a formatter can use to decide what to show, for diag events this is the diagnostics message.
sub summary { return "display this" }
```

So you just write a formatter that has a write($event, $number) method, that prints the right thing to STDERR/STDOUT when the event comes in. Loading the formatter is usually enough to make it be THE formatter.

**Comment by magnolia-k** [comment]
rewrote according to Test2::Formatter's API.

https://github.com/tokuhirom/Test-Pretty/pull/38
